lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,500 @@
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskId.Equals(_masterTaskId) && MasterTaskExists())
+            {
+                string msg = string.Format("Trying to add second master Task {0}.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _tasks.Add(taskId, new Tuple<TaskStateMachine, IConfiguration, IActiveContext>(new TaskStateMachine(), taskConfiguration, activeContext));
+
+            if (NumberOfTasks == _totalExpectedTasks && !MasterTaskExists())
+            {
+                Exceptions.Throw(new IMRUSystemException("There is no master task added."), Logger);
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of tasks in the task collection
+        /// </summary>
+        internal int NumberOfTasks
+        {
+            get { return _tasks.Count; }
+        }
+
+        /// <summary>
+        /// This method is called when receiving IRunningTask event during system SubmittingTasks state.
+        /// Adds the IRunningTask to the running tasks collection and update the task state to TaskRunning.
+        /// Throws IMRUSystemException if running tasks already contains this task or tasks collection doesn't contain this task.
+        /// </summary>
+        /// <param name="runningTask"></param>
+        internal void SetRunningTask(IRunningTask runningTask)
+        {
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already running.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (!_tasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _runningTasks.Add(runningTask.Id, runningTask);
+            UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+        }
+
+        /// <summary>
+        /// This method is called at the beginning of the recovery.
+        /// Clears the task collection, running task collection and resets the number of application error. 
+        /// </summary>
+        internal void Reset()
+        {
+            _tasks.Clear();
+            _runningTasks.Clear();
+            _numberOfAppError = 0;
+        }
+
+        /// <summary>
+        /// This method is called when receiving ICompletedTask event during task running or system shutting down.
+        /// Removes the task from running tasks
+        /// Changes the task state from RunningTask to CompletedTask
+        /// </summary>
+        /// <param name="taskId"></param>
+        internal void SetCompletedTask(string taskId)
```

Comment:
We only need id in this case. 
What we need to use in ICompletedTask is ActiveContext so that to dispose it. But all ActiveContexts have been in ActiveContextManager.